File: typescript/tests/beauty_board.test.ts
// beauty_board.test.ts
// Verifies the CircleAI.Beauty port: treatments, appointments-between,
// profiles, concern-based recommender.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryBeautyBoard,
  BeautyDomainContext,
  treatment,
  appointment,
  skinProfile,
} from "../src/beauty/index";

describe("InMemoryBeautyBoard", () => {
  it("adds treatments and returns appointments within a window, ordered", () => {
    const b = new InMemoryBeautyBoard();
    b.addTreatment(treatment("t1", "Facial", 60, 500, "ZAR"));
    assert.equal(b.getTreatment("t1")?.name, "Facial");
    b.book(appointment("a1", "Ann", "t1", new Date("2026-01-05T10:00:00Z"), null));
    b.book(appointment("a2", "Bea", "t1", new Date("2026-01-01T10:00:00Z"), "note"));
    b.book(appointment("a3", "Cy", "t1", new Date("2026-01-20T10:00:00Z"), null)); // out of window
    assert.deepEqual(
      b
        .appointmentsBetween(new Date("2026-01-01T00:00:00Z"), new Date("2026-01-10T00:00:00Z"))
        .map((a) => a.apptId),
      ["a2", "a1"],
    );
  });

  it("recommends treatments matching profile concerns, [] when no profile", () => {
    const b = new InMemoryBeautyBoard();
    b.addTreatment(treatment("t1", "Acne Clearing", 45, 400, "ZAR"));
    b.addTreatment(treatment("t2", "Anti-Ageing Serum", 30, 600, "ZAR"));
    b.addTreatment(treatment("t3", "Relaxing Massage", 60, 700, "ZAR"));
    assert.equal(b.recommendFor("nobody").length, 0);
    b.saveProfile(skinProfile("Ann", "oily", ["acne", "ageing"]));
    assert.deepEqual(
      b.recommendFor("Ann").map((t) => t.treatmentId).sort(),
      ["t1", "t2"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(BeautyDomainContext.systemPromptSnippet.includes("[DOMAIN: Beauty]"));
    assert.deepEqual(BeautyDomainContext.complianceFlags, ["POPIA", "Medicines_Act_cosmetic_claims"]);
    assert.deepEqual(BeautyDomainContext.suggestedTools, ["product_db", "ingredient_checker", "web_search"]);
  });
});

File: go/tests/beauty_board_test.go
// beauty_board_test.go
//
// Verifies the CircleAI.Beauty port (beauty_board.go): treatment add/get,
// appointment window query (ordered), profile save/get, and concern-based
// recommendation (case-insensitive, sorted).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestBeauty_TreatmentsAndAppointments(t *testing.T) {
	b := circleai.NewInMemoryBeautyBoard()
	b.AddTreatment(circleai.Treatment{TreatmentId: "t1", Name: "Acne Facial", DurationMinutes: 60, Price: circleai.DecimalFromInt(400), Currency: "ZAR"})
	if got, ok := b.GetTreatment("t1"); !ok || got.Name != "Acne Facial" {
		t.Fatalf("get treatment = %+v ok=%v", got, ok)
	}
	base := time.Date(2026, 7, 8, 9, 0, 0, 0, time.UTC)
	b.Book(circleai.Appointment{ApptId: "a2", ClientName: "Sam", TreatmentId: "t1", AtUtc: base.Add(3 * time.Hour)})
	b.Book(circleai.Appointment{ApptId: "a1", ClientName: "Sam", TreatmentId: "t1", AtUtc: base.Add(time.Hour)})
	b.Book(circleai.Appointment{ApptId: "a3", ClientName: "Sam", TreatmentId: "t1", AtUtc: base.Add(48 * time.Hour)})

	appts := b.AppointmentsBetween(base, base.Add(6*time.Hour))
	if len(appts) != 2 || appts[0].ApptId != "a1" || appts[1].ApptId != "a2" {
		t.Fatalf("appointments-between ordered failed: %+v", appts)
	}
}

func TestBeauty_Recommend(t *testing.T) {
	b := circleai.NewInMemoryBeautyBoard()
	b.AddTreatment(circleai.Treatment{TreatmentId: "t1", Name: "Acne Facial", Price: circleai.DecimalFromInt(400)})
	b.AddTreatment(circleai.Treatment{TreatmentId: "t2", Name: "Anti-Ageing Peel", Price: circleai.DecimalFromInt(700)})
	b.AddTreatment(circleai.Treatment{TreatmentId: "t3", Name: "Massage", Price: circleai.DecimalFromInt(500)})
	b.SaveProfile(circleai.SkinProfile{ClientName: "Sam", SkinType: "combination", Concerns: []string{"acne", "ageing"}})
	if got, ok := b.GetProfile("Sam"); !ok || got.SkinType != "combination" {
		t.Fatalf("get profile = %+v ok=%v", got, ok)
	}

	recs := b.RecommendFor("Sam")
	if len(recs) != 2 || recs[0].TreatmentId != "t1" || recs[1].TreatmentId != "t2" {
		t.Fatalf("recommend failed: %+v", recs)
	}
	if empty := b.RecommendFor("Unknown"); len(empty) != 0 {
		t.Fatalf("recommend for unknown client must be empty: %+v", empty)
	}
}
